docker/python3-scientific/create_zarr_optimus.py

```diff
@@ -0,0 +1,360 @@
+
+    gene_metric_values = []
+    for gene_id in gene_ids:
+        if gene_id in gene_id_to_metric_values:
+            gene_metric_values.append(gene_id_to_metric_values[gene_id])
+        else:
+            # if no metrics for a cell present in the count matrix then fill them with np.nans
+            gene_metric_values.append([np.nan] * ncols)
+
+    nrows = len(gene_ids)
+    if verbose:
+        logging.info("# of genes: {}".format(nrows))
+        logging.info("# of gene metadate metrics: {}".format(ncols))
+
+    # now insert the dataset that has the numeric values for the qc metrics for the genes
+    data_group.create_dataset(
+            "gene_metadata_numeric",
+            shape=(nrows, ncols),
+            compressor=COMPRESSOR,
+            dtype=np.float32,
+            chunks=(nrows, ncols),
+            data=gene_metric_values)
+
+
+def add_cell_metrics(data_group, input_path, cell_ids, verbose=False):
+    """Converts cell metrics from the Optimus pipeline to zarr file
+    Args:
+        data_group (zarr.hierarchy.Group): datagroup object for the zarr
+        input_path (str): file containing gene metrics name and values
+        cell_ids (list): list of cell ids
+        verbose (bool): whether to output verbose messages for debugging purposes
+    """
+
+    # read the gene metrics names and values
+    if input_path.endswith(".gz"):
+        with gzip.open(input_path, 'rt') as f:
+            cell_metrics = [row for row in csv.reader(f)]
+    else:
+        with open(input_path, 'r') as f:
+            cell_metrics = [row for row in csv.reader(f)]
+
+    # metric names for the cells
+    data_group.create_dataset(
+            "cell_metadata_numeric_name",
+            shape=(len(cell_metrics[0][1:]),),
+            compressor=COMPRESSOR,
+            dtype="<U40",
+            chunks=(len(cell_metrics[0][1:]),),
+            data=list(cell_metrics[0][1:]))
+
+    if verbose:
+        logging.info("# of cell_metadata_numeric_names: {}".format(len(cell_metrics[0][1:])))
+
+    cell_ids_location = {cell_id: index for index, cell_id in enumerate(cell_ids)}
+
+    cell_id_to_metric_values = {}
+    # ignore the first line with the cell metric names in text
+    ncols = 0
+    for row in cell_metrics[1:]:
+        # only consider cell_id that are also in the count_metrics
+        if not row[0] in cell_ids_location:
+            continue
+        row_values = []
+        for value_string in row[1:]:
+            # some of the standard deviation values do not exist for one reads matches
+            try:
+                value = np.float32(value_string)
+            except ValueError:
+                value = np.nan
+            row_values.append(value)
+        cell_id_to_metric_values[row[0]] = row_values
+
+        # note that all of these lengths are assumed to be equal and this check is already done in the pipeline
+        ncols = len(row_values)
```

Comment:
It seems this variable `ncols` will be overwritten in each loop, which means the final value will be the length of the `row_values ` in the last iteration, is this expected? Same question for `nrows` variable